### DISCUSSION TITLE
Passing file as cookie in the core.http runner

### DISCUSSION THREAD

**Discussion by zsmanjot** [StackStorm/st2#5631] [Packs]
We are using curl command in one of our projects. Since curl is not that secure so we thought to leverage the core.http method of stackstorm. But the issue is we are unable to pass the cookie using it. We have cookie stored in a file and it is a Netscape HTTP Cookie File. Do we have any method of how we can pass this file as cookie in the core.http runner.

**Comment by arm4b** [comment]
@zsmanjot Because you're asking a question, the initial st2 issue (https://github.com/StackStorm/st2/issues/5630) was converted to a Github Discussion here.
Please use the GH Discussions in the future and include the detailed information in the post description instead of a Title.
I moderated it for you.

**Comment by arm4b** [comment]
As for the question you're asking, per StackStorm docs (https://docs.stackstorm.com/reference/runners.html#id9) looks like http runner accepts `cookies` parameter so you can pass your custom cookies as a text instead of using a file.

**Comment by zsmanjot** [comment]
Hi @armab 

Thanks for replying. I tried using the that way. But it says me invalid credentials. May be the cookie is not passed correctly. We basically have a  Netscape HTTP Cookie File. This file is generated by libcurl. Here is the reference : https://curl.haxx.se/docs/http-cookies.html

 Via curl command we can use -b option to specify the file path. But with core.http we donot have such option i suppose. Let me know your views on it.

**Reply by arm4b** [reply]
Try to debug it one level deeper, like comparing the HTTP request headers that are sent when using the `curl` command and mimicking the same request with `core.http`.

**Reply by zsmanjot** [reply]
With curl command everything works fine. We are just stuck at cookies part though.

We are further checking it. But i believe there should be an option to pass file path in cookies like we do in curl command.

**Reply by arm4b** [reply]
Sure, because StackStorm is an Open Source project, you could contribute the new functionality to `core.http` runner to allow "file" as alternative input.

But I believe it would be easier to understand the existing `core.http` interface so you can technically send the same HTTP request.

**Reply by zsmanjot** [reply]
I agree. core.http provides an option to pass cookie as an object. But in my case since i mentioned that i am using Netscape cookie file format, so , it stores one cookie per physical line in the file with a bunch of associated meta data, each field separated with TAB. That file is called the cookiejar in curl terminology. The cookie file format is text based and stores one cookie per line. In this case  to pass this to curl or any other command is through a file. I was looking the same thing here in st2 through core.http runner. But now i think i have to make python/bash script to use curl only and then to use that script in the workflow.

I hope that we have this feature going forward where we can mention the file name that is read as a cookie by the runner. It will be really helpful for people who might use some  tool like Netscape for generating cookie .

**Reply by arm4b** [reply]
For the community record, could you perhaps provide an example of the cookie data you're trying to send?
Make sure to mask the secrets if any.

**Reply by zsmanjot** [reply]
@armab https://www.osec.doc.gov/webresources/p3p_user_admin_files/textmostly/slide11.html  This is an example set. I have the same kind of file.

**Reply by zsmanjot** [reply]
lLet me know your views on this.

**Comment by rush-skills** [comment]
Not sure if it helps, but can you try first using an action to read the file into a variable and then pass that into the cookies of the http runnner?

**Reply by zsmanjot** [reply]
I will try this one too. But keeping it simple and accessible should also be the focus here. Why not to leverage the parameters of the runner only, we just need to see how one can pass this file and this whole file can be read as a cookie.

@rush-skills  i will try your suggestion too and will update.

**Reply by zsmanjot** [reply]
Hi @armab @rush-skills 
Just to update core.http module is taking cookie appropriately now. Earlier we were passing it as an object separated by colon, but when we tried using "=" it started picking it up.

**Reply by arm4b** [reply]
Thanks for the follow-up.
Glad it worked!
